### PR TITLE
Guidelines: sample tests

### PR DESCRIPTION
**Pull request by hobovsky**
Discussion: #146 
Deploy preview: [here](https://deploy-preview-175--reverent-edison-2864ea.netlify.app/authoring/guidelines/sample-tests/)

### CODE REVIEW

File: content/authoring/guidelines/sample-tests.md [hunk 1]
```diff
@@ -7,4 +7,23 @@ next: /authoring/guidelines/submission-tests/
 
 # Writing Sample Tests
 
-_This page is a stub._
+Sample tests are a part of every kata which is often very underappreciated by both kata authors, and users attempting it. Bad sample tests can cause a lot of frustration, while a good suite of them often increases overall satisfaction of users from solving the kata. To make sample tests as good as possible, follow guidelines collected below.
+
+- **Conform to [General Coding Guidelines][authoring-guidelines-general-coding]**: tests are code too and as such should keep up to code quality standards.
```
Comment (kazk) [inline content/authoring/guidelines/sample-tests.md:12]:
```suggestion
- **Conform to [General Coding Guidelines](/authoring/guidelines/coding/)**: tests are code too and as such should keep up to code quality standards.
```

or add `[authoring-guidelines-general-coding]` at the bottom of the file.

```markdown
[authoring-guidelines-general-coding]: /authoring/guidelines/coding/
```

I prefer the latter because it's easier to maintain and read raw Markdown.

**Review by kazk — CHANGES_REQUESTED** [review]
We should split the "no sample tests" as an exception. There are multiple bullet points mentioning it. It's more clear to have a section dedicated for it. Clarify it's an exception, when it's acceptable, and how to minimize the risk of bad user experience.

File: content/authoring/guidelines/sample-tests.md (hunk 1 above)
Comment (hobovsky) [reply content/authoring/guidelines/sample-tests.md:12]: I remembered to add the reference at the bottom, and then forgot 🤦 

Added now.

File: content/authoring/guidelines/sample-tests.md [hunk 2]
```diff
@@ -9,27 +9,27 @@ next: /authoring/guidelines/submission-tests/
 
 Sample tests are a part of every kata which is often very underappreciated by both kata authors, and users attempting it. Bad sample tests can cause a lot of frustration, while a good suite of them often increases overall satisfaction of users from solving the kata. To make sample tests as good as possible, follow guidelines collected below.
```
Comment (DonaldKellett) [inline content/authoring/guidelines/sample-tests.md:10]:
```suggestion
Sample tests are a part of every kata which is often underappreciated by both kata authors and solvers. Bad sample tests can cause a lot of frustration, while a good suite of them often increases overall satisfaction of users from solving the kata. To make sample tests as good as possible, follow the guidelines collected below.
```

File: content/authoring/guidelines/sample-tests.md [hunk 3]
```diff
@@ -9,27 +9,27 @@ next: /authoring/guidelines/submission-tests/
...
+- **Conform to [General Coding Guidelines][authoring-guidelines-general-coding]**: tests should adhere to code quality standards as much as any other provided code.
+- Sample tests are **not a replacement for a good description.** The description alone should be enough to enable the user to complete the kata. Sample tests should be treated as supporting the description and never be the only source of information necessary to complete the kata.
 - **Avoid things like _"See sample tests for more examples"_**. To see the sample tests, users have to enter the kata trainer, which marks the kata as unfinished in their profile.
 - Sample tests should **include examples described in the description.**
 - **The test suite should be organized**, splitting the tests into different groups and subgroups. Each test framework provides tools to do that, generally with a possibility to give meaningful names to each (sub)group.
 - Sample tests can be used as additional examples for a couple of aspects:
   - In some languages, for example C, they serve as a good example on how solution function is called, how input and output parameters are passed to it, how memory management is handled, etc.
```
Comment (DonaldKellett) [inline content/authoring/guidelines/sample-tests.md:18]:
```suggestion
  - In some languages, for example C, they serve as a good example on how the solution function is called, how input and output parameters are passed to it, how memory management is handled, etc.
```

File: content/authoring/guidelines/sample-tests.md [hunk 4]
```diff
@@ -9,27 +9,27 @@ next: /authoring/guidelines/submission-tests/
...
   - Sample tests, in addition to the kata description, can clarify handling of edge cases, inputs which should be handled in some exceptional way, or errors.
-  - Test cases provided by author show how to use testing framework and assertion libraries to the users who'd like to add their own tests.
-  - One or two test cases with large inputs can serve as a hint on performance requirements, so users can see early that their solution is too slow or times out.  They can be initially commented out or disabled in some other way, with a note that they can be enabled if user wants to verify performance of their solution.
+  - Test cases provided by the author show how to use the testing framework and assertion libraries to the users who'd like to add their own tests.
+  - One or two test cases with large inputs can serve as a hint on performance requirements, enabling users to check that their solution meets efficiency standards before submitting. These tests can be initially disabled with a note that they can be enabled if user wants to verify performance of their solution.
```
Comment (DonaldKellett) [inline content/authoring/guidelines/sample-tests.md:21]:
```suggestion
  - One or two test cases with large inputs can serve as a hint on performance requirements, enabling users to check that their solution meets efficiency standards before submitting. These tests can be initially disabled with a note that they can be enabled if the user wants to verify the performance of their solution.
```

File: content/authoring/guidelines/sample-tests.md [hunk 5]
```diff
@@ -9,27 +9,27 @@ next: /authoring/guidelines/submission-tests/
...
 - **Sample tests should be fixed, predictable, and repeatable.** Avoid sample tests which are random or change between the runs.
-- Initial tests should compile correctly and it should be possible to run them, but they are allowed to fail.
+- Initial tests should compile and execute fully without crashes or other non-test suite errors. Initially, users should expect clear, test suite-generated assertion failures for incomplete deliverables described in the instructions.
 - **Test cases from sample tests should appear also in submission tests**, possibly as a part of fixed tests.
```
Comment (DonaldKellett) [inline content/authoring/guidelines/sample-tests.md:24]:
```suggestion
- **Assertions from sample tests should appear in submission tests as well**, possibly as a part of fixed tests.
```

File: content/authoring/guidelines/sample-tests.md [hunk 6]
```diff
@@ -9,27 +9,27 @@ next: /authoring/guidelines/submission-tests/
...
-- If possible, **sample tests should be easy to reproduce in user's local environment.** Many users work on their solutions with their local IDE, and it's very helpful for them to have all required information available. Preloaded snippet is especially problematic for them because it's not directly visible and it makes it difficult to reproduce sample tests locally. If your sample tests rely on anything from preloaded code, consider relevant guidelines related to [description][authoring-guidelines-description] and [preloaded code][authoring-guidelines-preloaded].
+- If possible, **sample tests should be easy to reproduce in the user's local environment.** Many users work on their solutions in their local IDE and it's very helpful for them to have all required information available. The preloaded code file is especially problematic for them because it's not directly visible which makes it difficult to reproduce sample tests locally. If your sample tests rely on anything from preloaded code, consider relevant guidelines related to [description][authoring-guidelines-description] and [preloaded code][authoring-guidelines-preloaded].
```
Comment (DonaldKellett) [inline content/authoring/guidelines/sample-tests.md:25]:
```suggestion
- If possible, **sample tests should be easy to reproduce in the user's local environment.** Many users work on their solutions in their local IDE and it's very helpful for them to have all the required information available. The preloaded code file is especially problematic for them because it's not directly visible which makes it difficult to reproduce sample tests locally. If your sample tests rely on anything from preloaded code, consider relevant guidelines related to the [description][authoring-guidelines-description] and [preloaded code][authoring-guidelines-preloaded].
```

**Review by DonaldKellett — CHANGES_REQUESTED** [review]

**Review by DonaldKellett — APPROVED** [review]

**Review by kazk — APPROVED** [review]
